src/pages/VisiumPerm.tsx

```diff
@@ -54,11 +55,29 @@ export default function VisiumPerm() {
       }
     });
   }, []);
-  const [current, send] = useMachine(formMachine);
 
+  const [current, send] = useMachine(formMachine);
+  const [warnBeforeSave, setWarnBeforeSave] = React.useState(false);
   const { serverError } = current.context;
 
   const onSubmit = async (values: RecordPermRequest) => {
+    const isLabwareStained = await stanCore
+      .GetLabwareOperations({
+        barcode: values.barcode,
+        operationType: 'Stain'
+      })
+      .then((r) => r.labwareOperations?.length);
+
+    if (!!isLabwareStained) {
+      handleSave(values);
+    } else {
+      setWarnBeforeSave(true);
+    }
```

Comment:
Just a question - Would it be more appropriate to do it as an xState service while scanning the labware itself than doing it on 'Save operation'?

Reply:
hmm good question. I guess that could be done by showing the pop up modal after the labware has been scanned if it doesnt have stains. I think i did it this way because in other pages like slotCopy(visium cDna) the warning modal shows up on submit

Reply:
It's ok. We can go with this as it is doing it like other pages 👍